src/components/views/dialogs/ElementProDialog.tsx

```diff
@@ -0,0 +1,157 @@
+                        },
+                    },
+                );
+                break;
+            case PostmessageAction.SetupComplete:
+                // Set as completed but let the user close the modal themselves
+                // so they have time to finish reading any information
+                this.setState({
+                    completed: true,
+                });
+                break;
+            case PostmessageAction.CloseDialog:
+                this.onFinished(true);
+                break;
+        }
+    }
+
+    private onFinished = (result: boolean) => {
+        if (result || this.state.completed) {
+            // We're done, close
+            this.props.requestClose();
+        } else {
+            Modal.createDialog(
+                ElementProConfirmCloseDialog,
+                {
+                    onFinished: result => {
+                        if (result) {
+                            this.props.requestClose();
+                        }
+                    },
+                },
+            );
+        }
+    }
+
+    private sendMessage = (message: IPostmessageResponseData) => {
+        this.iframeRef.current.contentWindow.postMessage(message, this.elementProSetupUrl);
+    }
+
+    private async sendAccountDetails() {
+        const openIdToken = await MatrixClientPeg.get().getOpenIdToken();
+        if (!openIdToken || !openIdToken.access_token) {
+            this.setState({
+                completed: true,
+                error: _t("Failed to connect to your homeserver. Please close this dialog and try again."),
+            });
+            return;
+        }
+        this.sendMessage({
+            action: PostmessageAction.ElementProAccountDetails,
+            account: {
+                accessToken: await MatrixClientPeg.get().getAccessToken(),
+                name: OwnProfileStore.instance.displayName,
+                openIdToken: openIdToken.access_token,
+                serverName: await MatrixClientPeg.get().getDomain(),
+                userLocalpart: await MatrixClientPeg.get().getUserIdLocalpart(),
+            },
+        });
+    }
+
+    public componentDidMount() {
+        window.addEventListener("message", this.messageHandler);
+    }
+
+    public componentWillUnmount() {
+        window.removeEventListener("message", this.messageHandler);
+    }
+
+    public render(): React.ReactNode {
+        return (
+            <BaseDialog
+                className="mx_ElementProBaseDialog"
+                onFinished={this.onFinished}
+                title={_t("Set up your own personal Element host")}
+                hasCancel={true}
+                fixedWidth={false}
+            >
+                <div className="mx_ElementProDialog_container">
+                    <iframe
+                        src={this.elementProSetupUrl}
+                        ref={this.iframeRef}
+                        sandbox="allow-forms allow-scripts allow-same-origin"
```

Comment:
it still seems really dangerous to me that we need `allow-same-origin` - a comment explaining why would be greatly appreciated.